server/schedulers/balance_region.go

```diff
@@ -121,63 +105,50 @@ func (s *balanceRegionScheduler) IsScheduleAllowed(cluster opt.Cluster) bool {
 	stores := cluster.GetStores()
-
-	// source is the store with highest region score in the list that can be selected as balance source.
-	f := s.hitsCounter.buildSourceFilter(s.GetName(), cluster)
-	source := s.selector.SelectSource(cluster, stores, f)
-	if source == nil {
-		schedulerCounter.WithLabelValues(s.GetName(), "no-source-store").Inc()
-		// Unlike the balanceLeaderScheduler, we don't need to clear the taintCache
-		// here. Because normally region score won't change rapidly, and the region
-		// balance requires lower sensitivity compare to leader balance.
-		return nil
-	}
-
-	sourceID := source.GetID()
-	log.Debug("store has the max region score", zap.String("scheduler", s.GetName()), zap.Uint64("store-id", sourceID))
-	sourceAddress := source.GetAddress()
-	sourceLabel := strconv.FormatUint(sourceID, 10)
-	s.counter.WithLabelValues("source-store", sourceAddress, sourceLabel).Inc()
-
-	for i := 0; i < balanceRegionRetryLimit; i++ {
-		// Priority picks the region that has a pending peer.
-		// Pending region may means the disk is overload, remove the pending region firstly.
-		region := cluster.RandPendingRegion(sourceID, core.HealthRegionAllowPending())
-		if region == nil {
-			// Then picks the region that has a follower in the source store.
-			region = cluster.RandFollowerRegion(sourceID, core.HealthRegion())
-		}
-		if region == nil {
-			// Last, picks the region has the leader in the source store.
-			region = cluster.RandLeaderRegion(sourceID, core.HealthRegion())
-		}
-		if region == nil {
-			schedulerCounter.WithLabelValues(s.GetName(), "no-region").Inc()
-			s.hitsCounter.put(source, nil)
-			continue
-		}
-		log.Debug("select region", zap.String("scheduler", s.GetName()), zap.Uint64("region-id", region.GetID()))
-
-		// We don't schedule region with abnormal number of replicas.
-		if len(region.GetPeers()) != cluster.GetMaxReplicas() {
-			log.Debug("region has abnormal replica count", zap.String("scheduler", s.GetName()), zap.Uint64("region-id", region.GetID()))
-			schedulerCounter.WithLabelValues(s.GetName(), "abnormal-replica").Inc()
-			s.hitsCounter.put(source, nil)
-			continue
-		}
-
-		// Skip hot regions.
-		if cluster.IsRegionHot(region) {
-			log.Debug("region is hot", zap.String("scheduler", s.GetName()), zap.Uint64("region-id", region.GetID()))
-			schedulerCounter.WithLabelValues(s.GetName(), "region-hot").Inc()
-			s.hitsCounter.put(source, nil)
-			continue
-		}
-
-		oldPeer := region.GetStorePeer(sourceID)
-		if op := s.transferPeer(cluster, region, oldPeer); op != nil {
-			schedulerCounter.WithLabelValues(s.GetName(), "new-operator").Inc()
-			return []*operator.Operator{op}
+	stores = filter.SelectSourceStores(stores, s.filters, cluster)
+	sort.Slice(stores, func(i, j int) bool {
+		return stores[i].RegionScore(cluster.GetHighSpaceRatio(), cluster.GetLowSpaceRatio(), 0) > stores[j].RegionScore(cluster.GetHighSpaceRatio(), cluster.GetLowSpaceRatio(), 0)
+	})
+	for _, source := range stores {
```

Comment:
If a store is frequently selected to cause hotspots, how will we handle them?

Reply:
The store needs to [pass the filters](https://github.com/pingcap/pd/pull/1823/files#diff-fe9ddab735e2ac8ca9604d89e0f10a5dR108). If a store has too many snapshots, or exceeds store limit, it will be ignored.

Reply:
the filter seem to only check target rather than source. If a store is selected source continuous, is it expected?